src/Evolution/Systems/GeneralizedHarmonic/Constraints.cpp

```diff
@@ -37,6 +37,63 @@ void three_index_constraint(
     *constraint_it = *d_spacetime_metric_it - *phi_it;
   }
 }
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+tnsr::a<DataType, SpatialDim, Frame> gauge_constraint(
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  auto constraint =
+      make_with_value<tnsr::a<DataType, SpatialDim, Frame>>(pi, 0.0);
+  gauge_constraint<SpatialDim, Frame, DataType>(
+      &constraint, gauge_function, spacetime_normal_one_form,
+      spacetime_normal_vector, inverse_spatial_metric,
+      inverse_spacetime_metric, pi, phi);
+  return constraint;
+}
+
+template <size_t SpatialDim, typename Frame, typename DataType>
+void gauge_constraint(
+    const gsl::not_null<tnsr::a<DataType, SpatialDim, Frame>*> constraint,
+    const tnsr::a<DataType, SpatialDim, Frame>& gauge_function,
+    const tnsr::a<DataType, SpatialDim, Frame>& spacetime_normal_one_form,
+    const tnsr::A<DataType, SpatialDim, Frame>& spacetime_normal_vector,
+    const tnsr::II<DataType, SpatialDim, Frame>& inverse_spatial_metric,
+    const tnsr::AA<DataType, SpatialDim, Frame>& inverse_spacetime_metric,
+    const tnsr::aa<DataType, SpatialDim, Frame>& pi,
+    const tnsr::iaa<DataType, SpatialDim, Frame>& phi) noexcept {
+  for (size_t a = 0; a < SpatialDim + 1; ++a) {
+    (*constraint).get(a) = gauge_function.get(a);
+    for (size_t i = 0; i < SpatialDim; ++i) {
+      for (size_t j = 0; j < SpatialDim; ++j) {
+        (*constraint).get(a) +=
+            inverse_spatial_metric.get(i, j) * phi.get(i, j + 1, a);
+      }
+    }
+    for (size_t b = 0; b < SpatialDim + 1; ++b) {
+      (*constraint).get(a) += spacetime_normal_vector.get(b) * pi.get(b, a);
+      for (size_t c = 0; c < SpatialDim + 1; ++c) {
```

Comment:
because all expressions summed over are symmetric in b and c, you can start c at b+1 (with appropriate factors of 2 in each expression) and add the diagonal term afterward for savings.